File: dongdaang/dynamic_programming/push_dominoes/Push_Dominoes.py
#https://leetcode.com/problems/push-dominoes/


class Solution:
    def pushDominoes(self, dominoes):
        n = len(dominoes)
        if n == 1:
            return dominoes

        force_point = []    #도미노를 넘어뜨린 지점만 담는 리스트
        res = []            #넘어뜨린 후 도미노 상태 담는 리스트
        ans = ''
        
        for i in range(n):
            res.append(dominoes[i])
            if dominoes[i] != '.':
                force_point.append(i)
        
        if not force_point:     #도미노를 넘어뜨린 지점이 없으면 도미노 그대로 반환
            return dominoes
        
        if dominoes[force_point[0]] == 'L':     #넘어뜨린 지점 중 가장 왼쪽에 있는 곳이 왼쪽으로 넘어진 경우
            for i in range(force_point[0]):
                res[i] = 'L'
        if dominoes[force_point[-1]] == 'R':        #넘어뜨린 지점 중 가장 오른쪽에 있는 곳이 오른쪽으로 넘어진 경우
            for i in range(n - 1, force_point[-1], -1):
                res[i] = 'R'
        
        for i in range(len(force_point) - 1):       #넘어뜨린 지점 사이의 경우의 수를 각각 처리
            if dominoes[force_point[i]] == 'L' and dominoes[force_point[i + 1]] == 'R':
                continue
            if dominoes[force_point[i]] == 'L' and dominoes[force_point[i + 1]] == 'L':
                for j in range(force_point[i] + 1, force_point[i + 1]):
                    res[j] = 'L'
            if dominoes[force_point[i]] == 'R' and dominoes[force_point[i + 1]] == 'R':
                for j in range(force_point[i] + 1, force_point[i + 1]):
                    res[j] = 'R'
            if dominoes[force_point[i]] == 'R' and dominoes[force_point[i + 1]] == 'L':
                tmp = force_point[i + 1] - force_point[i]
                m = tmp // 2
                if tmp % 2 == 0:
                    for j in range(force_point[i] + 1, force_point[i] + m):
                        res[j] = 'R'
                    for j in range(force_point[i + 1] - 1, force_point[i + 1] - m, -1):
                        res[j] = 'L'
                else:
                    for j in range(force_point[i] + 1, force_point[i] + m + 1):
                        res[j] = 'R'
                    for j in range(force_point[i + 1] - 1, force_point[i + 1] - m - 1, -1):
                        res[j] = 'L'
        
        for i in res:
            ans += i
            
        return ans
    
force_point = Solution()
print(force_point.pushDominoes('.L.R...LR..L..'))

File: CokeLee777/src/com/leetcode/dynamicprogramming/pushdominos/PushDominoes.java
package com.leetcode.dynamicprogramming.pushdominos;

public class PushDominoes {

    static class Solution{

        public String pushDominoes(String dominoes) {
            char[] a = dominoes.toCharArray();
            //투 포인터를 사용한 풀이
            for (int i = 0, L = -1, R = -1; i <= dominoes.length(); i++)
                //도미노의 끝점이거나 오른쪽으로 넘어지는 도미노라면
                if (i == a.length || a[i] == 'R') {
                    //마지막 끝 R이 L보다 인덱스가 크다면 모두 R로 치환
                    if (R > L)//R..R, turn all to R
                        while (R < i)
                            a[R++] = 'R';
                    R = i;
                    //도미노가 왼쪽으로 넘어지는 도미노라면
                } else if (a[i] == 'L')
                    //R이 아직 나오지 않았거나 마지막 L의 인덱스가 마지막 R보다 크다면
                    if (L > R || R == -1)//L..L, turn all to L
                        while (++L < i)
                            a[L] = 'L';
                    else { //R...L
                        L = i;
                        for (int lo = R + 1, hi = L - 1; lo < hi; ) {//one in the middle stays '.'
                            a[lo++] = 'R';
                            a[hi--] = 'L';
                        }
                    }
            return new String(a);
        }
    }

    static class Solution2{

        public String pushDominoes(String d) {
            d = 'L' + d + 'R';
            StringBuilder res = new StringBuilder();
            for (int i = 0, j = 1; j < d.length(); ++j) {
                if (d.charAt(j) == '.') continue;
                int middle = j - i - 1;
                if (i > 0)
                    res.append(d.charAt(i));
                if (d.charAt(i) == d.charAt(j))
                    for (int k = 0; k < middle; k++)
                        res.append(d.charAt(i));
                else if (d.charAt(i) == 'L' && d.charAt(j) == 'R')
                    for (int k = 0; k < middle; k++)
                        res.append('.');
                else {
                    for (int k = 0; k < middle / 2; k++)
                        res.append('R');
                    if (middle % 2 == 1)
                        res.append('.');
                    for (int k = 0; k < middle / 2; k++)
                        res.append('L');
                }
                i = j;
            }
            return res.toString();
        }
    }
}
